compiler/rustc_middle/src/ty/trait_def.rs

```diff
@@ -177,25 +189,65 @@ impl<'tcx> TyCtxt<'tcx> {
+        // `impls_for_ref_x`, while using `TreatParams::AsInfer` while actually adding them (in
+        // `trait_impls_of_provider`).
         //
         // This way, when searching for some impl for `T: Trait`, we do not look at any impls
         // whose outer level is not a parameter or projection. Especially for things like
         // `T: Clone` this is incredibly useful as we would otherwise look at all the impls
         // of `Clone` for `Option<T>`, `Vec<T>`, `ConcreteType` and so on.
-        if let Some(simp) = fast_reject::simplify_type(self, self_ty, TreatParams::AsPlaceholder) {
-            if let Some(impls) = impls.non_blanket_impls.get(&simp) {
+        if let TyKind::Ref(_, ref_ty, mutability) = self_ty.kind() {
+            // Before simplifying the referenced type, we start by checking impls for references to
+            // a non-simplifiable type.
+            if let Some(impls) =
+                impls.non_blanket_impls.get(&SimplifiedType::RefSimplifiedType(*mutability))
+            {
                 for &impl_def_id in impls {
                     if let result @ Some(_) = f(impl_def_id) {
                         return result;
                     }
                 }
             }
+
+            // Then we can check impls for references to a simplifiable type, when possible.
+            if let Some(simplified_ref_ty) =
+                fast_reject::simplify_type(self, *ref_ty, TreatParams::AsPlaceholder)
+            {
+                if let Some(impls) = impls.impls_for_ref_x.get(&simplified_ref_ty) {
+                    for &impl_def_id in impls {
+                        if let result @ Some(_) = f(impl_def_id) {
+                            return result;
+                        }
+                    }
+                }
+            } else {
+                for &impl_def_id in impls.impls_for_ref_x.values().flatten() {
+                    if let result @ Some(_) = f(impl_def_id) {
+                        return result;
+                    }
+                }
+            }
         } else {
-            for &impl_def_id in impls.non_blanket_impls.values().flatten() {
-                if let result @ Some(_) = f(impl_def_id) {
-                    return result;
+            if let Some(simplified_self_ty) =
+                fast_reject::simplify_type(self, self_ty, TreatParams::AsPlaceholder)
+            {
+                if let Some(impls) = impls.non_blanket_impls.get(&simplified_self_ty) {
+                    for &impl_def_id in impls {
+                        if let result @ Some(_) = f(impl_def_id) {
+                            return result;
+                        }
+                    }
+                }
+            } else {
+                for &impl_def_id in impls.non_blanket_impls.values().flatten() {
+                    if let result @ Some(_) = f(impl_def_id) {
+                        return result;
+                    }
                 }
             }
```

Comment:
the else branch is shared with `RefSimplifiedType`, isn't it? I think it makes sense to do
```rust 
if let Some(simp) = fast_reject::simplify_type(self, self_ty, TreatParams::AsPlaceholder) {
    impls.blanket_impls.iter()
     
    impls.non_blanket_impls.get(simp).iter()

    if let ty::Ref(_, ref_ty, mutablity) {
        if let Some(ref_simp) = fast_reject::simplify_type(self, *ref_ty, TreatParams::AsPlaceholder) {
            impls.impls_for_ref_x.get(ref_simp).iter()
        } else {
            impls.impls_for_ref_x.iter().flatten()
        }
    }
} else {
    self.all_impls(def_id)
}
```

Reply:
Good idea. I've followed this suggestion when rebasing.